File: android/src/main/java/com/facebook/sonar/testing/SonarConnectionMock.java
/*
 *  Copyright (c) 2018-present, Facebook, Inc.
 *
 *  This source code is licensed under the MIT license found in the LICENSE
 *  file in the root directory of this source tree.
 *
 */
package com.facebook.flipper.testing;

import com.facebook.flipper.core.SonarArray;
import com.facebook.flipper.core.SonarConnection;
import com.facebook.flipper.core.SonarObject;
import com.facebook.flipper.core.SonarReceiver;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class SonarConnectionMock implements SonarConnection {
  public final Map<String, SonarReceiver> receivers = new HashMap<>();
  public final Map<String, List<Object>> sent = new HashMap<>();

  @Override
  public void send(String method, SonarObject params) {
    final List<Object> paramList;
    if (sent.containsKey(method)) {
      paramList = sent.get(method);
    } else {
      paramList = new ArrayList<>();
      sent.put(method, paramList);
    }

    paramList.add(params);
  }

  @Override
  public void send(String method, SonarArray params) {
    final List<Object> paramList;
    if (sent.containsKey(method)) {
      paramList = sent.get(method);
    } else {
      paramList = new ArrayList<>();
      sent.put(method, paramList);
    }

    paramList.add(params);
  }

  @Override
  public void reportError(Throwable throwable) {}

  @Override
  public void receive(String method, SonarReceiver receiver) {
    receivers.put(method, receiver);
  }
}

File: iOS/FlipperKitTestUtils/SonarConnectionMock.m
/*
 *  Copyright (c) 2018-present, Facebook, Inc.
 *
 *  This source code is licensed under the MIT license found in the LICENSE
 *  file in the root directory of this source tree.
 *
 */
#import "SonarConnectionMock.h"

@implementation SonarConnectionMock

- (instancetype)init
{
  if (self = [super init]) {
    _connected = YES;
    _receivers = @{};
    _sent = @{};
  }
  return self;
}


- (void)send:(NSString *)method withParams:(NSDictionary *)params
{
  if (_connected) {
    NSMutableDictionary *newSent = [NSMutableDictionary new];
    [newSent addEntriesFromDictionary:_sent];
    if (newSent[method]) {
      newSent[method] = [_sent[method] arrayByAddingObject:params];
    } else {
      newSent[method] = @[params];
    }
    _sent = newSent;
  }
}

- (void)receive:(NSString *)method withBlock:(SonarReceiver)receiver
{
  if (_connected) {
    NSMutableDictionary *newReceivers = [NSMutableDictionary new];
    [newReceivers addEntriesFromDictionary:_receivers];
    newReceivers[method] = receiver;
    _receivers = newReceivers;
  }
}

@end
